Rehaul: Version control with Git (#1069)
appsmithorg/appsmith-docs: website/docs/advanced-concepts/version-control-with-git/README.mdx

:::

<VideoEmbed host="youtube" videoId="emdZaG-tzgw" title="Version Control with Git " caption="Version Control with Git "/>

Version Control works with any Git hosting service that supports **SSH protocol** and **deploy keys**. HTTPS Git connections are currently not supported on Appsmith.

## Connect to a Git repository

The first step to use Version control in Appsmith is to connect your remote repository with the app. Please follow the steps given below:

### Create a new repository

Create a new repository on your git service and copy the `SSH` URL. The repository should be empty except `README.md`, `.gitignore`, and `LICENSE` files (These can be auto-created when you create a new repository.)

### Generate a deploy key

1. Open the app you want to connect to git and click on the `Connect GIT` button on the bottom bar.
2. In the Git connection window, paste your repository’s `SSH` URL and click on `Generate key` to generate an `SSH` key.

>  <VideoEmbed host="youtube" videoId="oFfdIwhSPL8" title="Generating a Deploy key" caption="Generating a Deploy key"/>

Reviewer: replace with screenshot

Appsmith supports two types of keys -  **ECDSA 256** and **RSA 4096**. Please pick the one that's compatible with your Git hosting service. 

### Adding the deploy key in the Repository

> Copy the generated deploy key and add it to your repository. Please note that you have to **enable write access** for pushing the changes to the remote repository from Appsmith.

Reviewer: ```suggestion
Copy the `SSH key` generated in the preceding section. 
```

Removing the enable write access line as it is confusing there. Anyway we are showing how to do this for individual services

<Tabs queryString="current-edition">
<TabItem label="GitHub" value="github">

> * Open the settings of the repository and go to deploy keys.

Reviewer: ```suggestion
* In your repository, go to the **Settings** tab.
* On the navbar, under **Security**, click **Deploy Keys**. 
```

* Click **Add deploy key**. Add a title for the key and paste the generated key into the key section.

> * Check the **Allow write access** checkbox and add the key.

Reviewer: ```suggestion
* Select the **Allow write access** checkbox.
* Click **Add key**.
```

>   <VideoEmbed host="youtube" videoId="4xMHO4G2hEA" title="Adding the deploy key in the Repository | Github" caption="Adding the deploy key in the Repository | Github"/>

Reviewer: A screenshot will suffice

</TabItem>
<TabItem label="Gitlab" value="gitlab">

> 

Reviewer: Can we ensure the UI elements below in both GitLab and Bitbucket are styles as per Google style guide?

* Open the Settings> Repository and expand the deploy keys section.
* Click **Add key** Add the title and paste the generated key in the key section.
* Check the **Grant write permissions**" to this key and add the key.

>  <VideoEmbed host="youtube" videoId="9aaiE6OERW0" title="Adding the deploy key in the Repository | Gitlab" caption="Adding the deploy key in the Repository | Gitlab"/>

Reviewer: Replace with one screenshot

</TabItem>
<TabItem label="Bitbucket" value="bitbucket">

* Open **Personal settings** from the bottom right corner and go to **SSH keys**.
* Click **Add key**, add the label and paste the generated key in the key section.

>  <VideoEmbed host="youtube" videoId="A8ZOvW1CVIk" title="Adding the deploy key in the Repository | Bitbucket" caption="Adding the deploy key in the Repository | Bitbucket"/>

Reviewer: replace with screenshot

</TabItem>
</Tabs>

:::info

> **AWS CodeCommit** isn't supported as a source control repository due to their lack of support for deploy keys.

Reviewer: don't think we should such information in the documentation. Agree it might be a support question but calling it out here in this case doesn't look right.  Instead if we can have a general callout saying Appsmith supports only hosting services that provide support for deploy keys?

:::

> ### User configurations

Reviewer: ```suggestion
### User configuration
```

> * In user settings, define the user configurations. By default, the global user configurations are used. Click the **Edit** button or directly open the profile section to change the global configurations. If you want to add a user configuration specific to the current app, you must clear the **use default configuration** checkbox and enter the author details below.

Reviewer: there is some confusion here.  We are not communicating clearly to the user what global configuration means. They will try to find the Edit button in the git connect modal.  It would take them a while to figure that this on the workspace profile under Git user config.  Let's rewrite this para to communicate this clearly.

 <VideoEmbed host="youtube" videoId="d5R1MYKtpCM" title="User Settings | Git Sync" caption="User Settings | Git Sync"/>

> Finally, click on **connect**, and when the connection is successful, your repository will have a `README.md` file. On the Git window, you’ll move to the **Deploy** section, where you can make an initial commit.

Reviewer: this should be part of the step ordering right? Also, let's use number ordering everwhere where it is a procedure.  

And its not just `README.md` file that gets added to the repo. A bunch of stuff including two folders and a few JSON files also get added. 

Also unless they make any changes they don't have to initially commit anything.

Author: I think when the connection is successful, a README file gets added. An initial commit should be done to update the re[o the current state of the app. Currently, the repo is empty. The initial commit is a change itself as the repo does not have anything. The initial commit adds the app to the repository. If you don't make the initial commit, the repo will be empty even if the connection is successful.

Author: Sorry just checked. The initial commit step is automated now. Previously you had to make the initial commit manually.

<VideoEmbed host="youtube" videoId="Z67SfBpKrnk" title="Initial Commit | Git Sync" caption="Initial Commit | Git Sync"/>

> ## Updating local file path (optional)

Reviewer: ```suggestion
## Update local file path
```

> Appsmith clones the git repositories in the local filesystem, attached to the persistent volume within the docker container. To maintain the git repositories, you need a file path that points to the volume within the docker container. You can achieve this by just updating the relevant environment variable.

Reviewer: ```suggestion
For self-hosted instances, Appsmith clones the Git repositories in the local file system, attached to the persistent volume within the Docker container. To maintain the Git repositories, you need a file path that points to the volume within the Docker container.
```

> #### Custom Git root

Reviewer: remove this heading

> To point to a custom Git Root where the git repositories is persisted, update the `env` variable called `APPSMITH_GIT_ROOT` to point to your custom file path.

Reviewer: ```suggestion
To update the file path to point to a custom Git root where the repositories are persisted:

 1.  Update the `APPSMITH_GIT_ROOT` environment variable as shown below:
```

```

> APPSMITH_GIT_ROOT=./path/to/repo/directory 

Reviewer: ```suggestion
APPSMITH_GIT_ROOT=./<path-to-repo-directory>
```

```

> Please remember to restart the container to apply changes.

Reviewer: ```suggestion
2. Restart the container to apply the changes.
```

:::info

> If the file path isn't present, git repositories will be cloned, but this won't be persistent, and Appsmith will try to clone the repositories in case they got deleted by docker restart, etc.

Reviewer: This seems to be a but unclear. need to rewrite this.
